### ISSUE TITLE
UNKNOWN fs type and mount point fields in log files from Hopper system

### ISSUE THREAD

**Issue by shanedsnyder** [darshan-hpc/darshan#127]
In GitLab by @shanedsnyder on Sep 24, 2015, 16:26

The darshan-parser output for log files generated on Hopper will sometimes show UNKNOWN for the mount point and fs type.  This only happens for files that are _not_ opened on rank 0.  The problem does not appear in globally shared files, or files at rank 0.

Closer inspection of the log file indicates that the CP_DEVICE field is set to zero.  Preliminary test runs that print the CP_DEVICE field at the time that the file is opened indicate that it is correct, however.  Maybe the data is being corrupted when adjusting that field to match device IDs detected at rank 0?

Another observation is that this is occurring for files opened on the $SCRATCH (Lustre) file system on Hopper, which has an unusually large device ID of 3592384858.

Need to find a simpler test case and determine if this problem is present on Beagle as well.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Sep 24, 2015, 16:26

The initial description of this problem might be wrong.  I'm looking more closely at a GCRM example that reproduces the problem every time, and the issue _is_ occurring for shared files.  

The underlying problem in this case is that a file is opened with pnetcdf, but then is never opened (at least on some processes) with POSIX.  This is perhaps a lazy open optimization.

As a result, some processes have file records with incomplete fields (things like the device id are not filled in until posix open time).  Depending on how the reduction works, this can result in the final reduced file record having incomplete fields as well.

In general the code needs to be aware of these "incomplete" file records and handle them properly, especially when they are reduced with file records that are fully populated.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Sep 24, 2015, 16:26

Fixes for this problem are in trunk but untested.  The Hopper license server is down for maintenance right now.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Sep 24, 2015, 16:26

The fix seems correct for the reduction problem, but Hopper still reports UNKNOWN mount point and type for some files.  Now the actual output file does not have zero'd out data, but instead lists a device id (30) that does not show up in the mount table.

Need to collect more information across all of the processes while I have a test case that reproduces this variation of the problem.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Sep 24, 2015, 16:26

We discovered that the correct device id always appears when statting the parent directory of a file on Hopper.  As a result we have added a workaround that can be turned on by specifying --enable-st-dev-workaround at configure time.  If this option is set, then Darshan will still stat the file being opened to determine block size etc, but it will also stat the parent directory to retrieve the device id.

This has been committed to svn and tested on hopper.
